src/Simplex/Messaging/Agent.hs

```diff
@@ -255,31 +266,58 @@ processCommand c (connId, cmd) = case cmd of
 newConn :: AgentMonad m => AgentClient -> ConnId -> Maybe InvitationId -> Int -> m (ConnId, SMPQueueInfo)
 newConn c connId viaInv connLevel = do
   srv <- getSMPServer
-  (rq, qInfo) <- newReceiveQueue c srv
+  (rq, qInfo) <- newRcvQueue c srv
   g <- asks idsDrg
   let cData = ConnData {connId, viaInv, connLevel}
   connId' <- withStore $ \st -> createRcvConn st g cData rq
   addSubscription c rq connId'
   pure (connId', qInfo)
 
-joinConn :: forall m. AgentMonad m => AgentClient -> ConnId -> SMPQueueInfo -> ReplyMode -> Maybe InvitationId -> Int -> m ConnId
-joinConn c connId qInfo (ReplyMode replyMode) viaInv connLevel = do
-  (sq, senderKey, verifyKey) <- newSendQueue qInfo
+joinConn :: AgentMonad m => AgentClient -> ConnId -> SMPQueueInfo -> ConnInfo -> Maybe InvitationId -> Int -> m ConnId
+joinConn c connId qInfo cInfo viaInv connLevel = do
+  (sq, senderKey, verifyKey) <- newSndQueue qInfo
   g <- asks idsDrg
   let cData = ConnData {connId, viaInv, connLevel}
   connId' <- withStore $ \st -> createSndConn st g cData sq
-  connectToSendQueue c sq senderKey verifyKey
-  when (replyMode == On) $ createReplyQueue connId' sq
+  confirmQueue c sq senderKey cInfo
+  activateQueueJoining c connId' sq verifyKey 1 60
   pure connId'
+
+activateQueueJoining :: forall m. AgentMonad m => AgentClient -> ConnId -> SndQueue -> VerificationKey -> Int -> Int -> m ()
+activateQueueJoining c connId sndQ verifyKey initialDelaySec increaseDelayAfterSec = do
+  let onActivationCallback = removeActivation c connId >> cleanupSignatureKey sndQ >> createReplyQueue connId sndQ
+  a <- async $ activateQueue c sndQ verifyKey initialDelaySec increaseDelayAfterSec onActivationCallback
+  addActivation c connId a
   where
     createReplyQueue :: ConnId -> SndQueue -> m ()
     createReplyQueue cId sq = do
       srv <- getSMPServer
-      (rq, qInfo') <- newReceiveQueue c srv
+      (rq, qInfo') <- newRcvQueue c srv
       addSubscription c rq cId
       withStore $ \st -> upgradeSndConnToDuplex st cId rq
       sendControlMessage c sq $ REPLY qInfo'
 
+letConf :: AgentMonad m => AgentClient -> ConnId -> ConfirmationId -> ConnInfo -> m ConnId
+letConf c connId confirmationId ownConnInfo =
+  withStore (`getConn` connId) >>= \case
+    SomeConn SCRcv (RcvConnection _ rq) -> processConfirmation' rq
+    _ -> throwError $ CMD PROHIBITED
+  where
+    processConfirmation' :: AgentMonad m => RcvQueue -> m ConnId
+    processConfirmation' rq = do
+      withStore $ \st -> approveConfirmation st confirmationId ownConnInfo
+      confirmation <- withStore $ \st -> getConfirmation st confirmationId
+      let sndKey = senderKey (confirmation :: Confirmation)
+      processConfirmation c rq sndKey
+      withStore $ \st -> removeApprovedConfirmation st connId
+      pure connId
+
+processConfirmation :: AgentMonad m => AgentClient -> RcvQueue -> SenderPublicKey -> m ()
+processConfirmation c rq sndKey = do
+  withStore $ \st -> setRcvQueueStatus st rq Confirmed
```

Comment:
the queue should be marked as confirmed when SMP confirmation arrives, not when it is LET by the initiating party

Reply:
although, given that strictly speaking there can be several confirmation, maybe it's ok...